### PR TITLE
refactor: converted ResultSet to functional component

### PR DESCRIPTION
**Pull request by EugeneTorap**
### SUMMARY
Converted ResultSet class component into functional component.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API
@ad-m @AAfghahi @geido @lyndsiWilliams ^^

### CODE REVIEW

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx [hunk 1]
```diff
@@ -173,106 +160,93 @@ const updateDataset = async (
...
+      prevQuery.results.data.length > 0
     ) {
-      this.setState({ data: nextProps.query.results.data }, () =>
-        this.clearQueryResults(nextProps.query),
-      );
+      // this.setState({ data: prevQuery.results.data }, () =>
+      //   clearQueryResults(prevQuery),
+      // );
+      setCachedData(prevQuery.results.data);
+      actions.clearQueryResults(prevQuery);
```
Comment (EugeneTorap) [inline superset-frontend/src/SqlLab/components/ResultSet/index.tsx:222]: @lyndsiWilliams Do we have `useStateWithCallback` for it?

File: superset-frontend/src/SqlLab/components/ResultSet/ResultSet.test.jsx [hunk 2]
```diff
@@ -105,51 +105,51 @@ test('renders a Table', () => {
   expect(wrapper.find(FilterableTable)).toExist();
 });
 
-describe('componentDidMount', () => {
```
Comment (EugeneTorap) [inline superset-frontend/src/SqlLab/components/ResultSet/ResultSet.test.jsx:108]: @lyndsiWilliams How should I rewrite these tests? or just remove it?

File: superset-frontend/src/SqlLab/components/ResultSet/ResultSet.test.jsx [hunk 3]
```diff
@@ -177,7 +179,7 @@ test('should render empty results', () => {
 test('should render cached query', () => {
   const wrapper = shallow(<ResultSet {...cachedQueryProps} />);
   const cachedData = [{ col1: 'a', col2: 'b' }];
-  wrapper.setState({ data: cachedData });
+  wrapper.setState({ data: cachedData }); // How to set state to Functional Component?
```
Comment (EugeneTorap) [inline superset-frontend/src/SqlLab/components/ResultSet/ResultSet.test.jsx:182]: @lyndsiWilliams How to set state to Functional Component?

**Comment by AAfghahi** [comment]
Hey! Thank you for this, though what's the progress like on the four other tickets that you have open?

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx [hunk 4]
```diff
@@ -671,180 +607,172 @@ export default class ResultSet extends React.PureComponent<
         )}
       </ReturnedRows>
     );
+  };
+
+  let sql;
```
Comment (AAfghahi) [inline superset-frontend/src/SqlLab/components/ResultSet/index.tsx:612]: hmm i would have to think about this more, but I wonder if this being in a function is better practice. Either you can put it in the return. Which feels clunky, or maybe create another function for it. I'd have to think through what the differences are.

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx [hunk 5]
```diff
@@ -100,20 +100,6 @@ interface ResultSetProps {
   defaultQueryLimit: number;
 }
 
-interface ResultSetState {
```
[the remark this answers is not in the document]
Comment (EugeneTorap) [reply superset-frontend/src/SqlLab/components/ResultSet/index.tsx:103]: I've moved the types to useState<...>
For example: `useState<DatasetOptionAutocomplete[]>([]);`
Is it OK?

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx [hunk 6]
```diff
@@ -428,76 +390,53 @@ export default class ResultSet extends React.PureComponent<
...
+    actions.reFetchQueryResults(query);
+  };
 
-  renderControls() {
-    if (this.props.search || this.props.visualize || this.props.csv) {
-      let { data } = this.props.query.results;
-      if (this.props.cache && this.props.query.cached) {
-        ({ data } = this.state);
+  const renderControls = () => {
+    const saveModalAutocompleteValue = '';
```
[the remark this answers is not in the document]
Comment (EugeneTorap) [reply superset-frontend/src/SqlLab/components/ResultSet/index.tsx:432]: @AAfghahi The value is static and typescript show a error for `setSaveModalAutocompleteValue`.
What should I do?

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx (hunk 4 above)
Comment (EugeneTorap) [reply superset-frontend/src/SqlLab/components/ResultSet/index.tsx:612]: @AAfghahi Let's refactor it in next PR

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx (hunk 1 above)
Comment (lyndsiWilliams) [reply superset-frontend/src/SqlLab/components/ResultSet/index.tsx:222]: Are you referring to setting the state on lines 218-220? You might be able to use the [useCallback](https://reactjs.org/docs/hooks-reference.html#usecallback) hook for this.

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (lyndsiWilliams) [reply superset-frontend/src/SqlLab/components/ResultSet/index.tsx:432]: What error was it giving you?

File: superset-frontend/src/SqlLab/components/ResultSet/ResultSet.test.jsx (hunk 2 above)
Comment (lyndsiWilliams) [reply superset-frontend/src/SqlLab/components/ResultSet/ResultSet.test.jsx:108]: You can test the functionality that was done within these tests, they were converted to functional but still live on these lines: https://github.com/apache/superset/blob/4ec10c23014cb9e5fe2cedf6407bb711cf05dbb8/superset-frontend/src/SqlLab/components/ResultSet/index.tsx#L192-L227

File: superset-frontend/src/SqlLab/components/ResultSet/ResultSet.test.jsx (hunk 3 above)
Comment (lyndsiWilliams) [reply superset-frontend/src/SqlLab/components/ResultSet/ResultSet.test.jsx:182]: I believe this is setting the `query.results.data` as seen in `newProps` here:
https://github.com/apache/superset/blob/f62aa7b1b8cddd6f2655dbb657b730b5e27778df/superset-frontend/src/SqlLab/components/ResultSet/ResultSet.test.jsx#L88-L96
so you'll need to change the render's props in RTL to reflect that.

**Pull request closed without merging**